### PR TITLE
Add metadata for testcontainers:1.19.8

### LINKED ISSUE #492: com.github.dockerjava.api.model.RuntimeInfo
please add `com.github.dockerjava.api.model.RuntimeInfo` to the reachability metadata for testcontainers. without it, a trivial program launching a container doesnt run as a graalvm native image

### PR DESCRIPTION
**Pull request by Kehrlann**
## What does this PR do?

Adds support for Testcontainers `1.19.8` and up, used by the current Spring Boot `3.3.x` line. Reuses #132 + #301 and upgrades the version.

Closes #492.

## Code sections where the PR accesses files, network, docker or some external service

Tests are not enabled, same as current tests for Testcontainers `1.17.6` because it pulls a docker image with critical CVEs.

## Checklist before merging
- [x] I have considered including reachability metadata directly in the library or the framework (see [our contributing document](https://github.com/oracle/graalvm-reachability-metadata/blob/master/CONTRIBUTING.md))
- [x] I am the original author of all content provided in the pull request, and I did not copy the content from any other source (see [the tests section from our contributing document](https://github.com/oracle/graalvm-reachability-metadata/blob/master/CONTRIBUTING.md#tests))
- [x] For all tests where I am not the sole author, I have added a comment that proves I may publish them under the specified license (see [the tests section from our contributing document](https://github.com/oracle/graalvm-reachability-metadata/blob/master/CONTRIBUTING.md#tests))
- [x] I have properly formatted metadata files (see [our formatting guide](https://github.com/oracle/graalvm-reachability-metadata/blob/master/CONTRIBUTING.md#format-metadata-files))
- [x] I have added thorough tests (see [the tests section from our contributing document](https://github.com/oracle/graalvm-reachability-metadata/blob/master/CONTRIBUTING.md#Tests))
- [x] I have filled all places where my pull request accesses files, network, docker, or any other external service (see the section above)

### CODE REVIEW

**Comment by alina-yur** [comment]
thank you, @Kehrlann!❤️

File: metadata/org.testcontainers/testcontainers/1.19.8/resource-config.json [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+                }
+            },
+            {
+                "pattern": "\\Qcom/github/dockerjava/zerodep/shaded/org/apache/hc/client5/version.properties\\E",
+                "condition": {
+                    "typeReachable": "com.github.dockerjava.zerodep.shaded.org.apache.hc.core5.util.VersionInfo"
+                }
+            },
+            {
+                "pattern": "\\Qorg/slf4j/impl/StaticLoggerBinder.class\\E",
```
Comment (dnestoro) [inline metadata/org.testcontainers/testcontainers/1.19.8/resource-config.json:17]: Why is this class added as a resource? Is it necessary?

File: tests/src/org.testcontainers/testcontainers/1.19.8/build.gradle [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    implementation "org.testcontainers:testcontainers:$libraryVersion"
+    runtimeOnly "org.slf4j:slf4j-simple:1.7.36"
+    testImplementation 'org.assertj:assertj-core:3.22.0'
+}
+
+graalvmNative {
+    binaries {
+        test {
+            // We have to manually add the metadata for JNA, as testcontainers use JNA
+            configurationFileDirectories.from(tck.metadataRoot.dir('net.java.dev.jna/jna/5.8.0'))
```
Comment (dnestoro) [inline tests/src/org.testcontainers/testcontainers/1.19.8/build.gradle:25]: Have you tried to add JNA metadata in `META-INF/native-image` directory of the test? Check [this PR](https://github.com/oracle/graalvm-reachability-metadata/pull/257/files#diff-9957db8c55035c5fd697f91c7a7a682f694549386d2fb4117ddd42f995e0af95) for example. That way your tests will have additional metadata, but only for their run. Metadata provided like this won't be part of the metadata we are shipping in the release, and will be used only for test purposes.

More examples of such a usages can be found in:
https://github.com/oracle/graalvm-reachability-metadata/pull/118
https://github.com/oracle/graalvm-reachability-metadata/pull/141
https://github.com/oracle/graalvm-reachability-metadata/pull/397

File: metadata/org.testcontainers/testcontainers/1.19.8/reflect-config.json [hunk 3]
```diff
@@ -0,0 +1,4219 @@
+[
```
Comment (dnestoro) [inline metadata/org.testcontainers/testcontainers/1.19.8/reflect-config.json:1]: This is pretty huge `reflect-config.json`. Are you sure that all of these entries are necessary? For example there are lots of entries like this:
```
{
        "name": "java.util.LinkedHashMap",
        "queryAllDeclaredMethods": true,
        "condition": {
            "typeReachable": "org.testcontainers.dockerclient.DockerClientProviderStrategy"
        },
        "queryAllDeclaredConstructors": true
    },
    {
        "name": "java.util.LinkedHashMap",
        "condition": {
            "typeReachable": "org.testcontainers.dockerclient.EnvironmentAndSystemPropertyClientProviderStrategy"
        }
    },
    {
        "name": "java.util.LinkedHashMap",
        "condition": {
            "typeReachable": "org.testcontainers.dockerclient.TestcontainersHostPropertyClientProviderStrategy"
        }
    },
    {
        "name": "java.util.LinkedHashMap",
        "queryAllDeclaredMethods": true,
        "condition": {
            "typeReachable": "org.testcontainers.shaded.com.fasterxml.jackson.databind.ObjectMapper"
        },
        "queryAllDeclaredConstructors": true
    },
```

or something like 
```
  {
        "name": "java.net.UnixDomainSocketAddress",
        "condition": {
            "typeReachable": "org.testcontainers.dockerclient.DockerClientProviderStrategy"
        },
        "methods": [
            {
                "name": "of",
                "parameterTypes": [
                    "java.lang.String"
                ]
            }
        ]
    },
```

do we really need to include all of this stuff? Please take a quick look whether there are entries that can be excluded or not? If you find such entries, please update your [user-code-filter.json](https://github.com/oracle/graalvm-reachability-metadata/pull/533/files#diff-16ce117efd6cc1e3d7b2a4e2d8adb3e30d6af5628baede0fb46a81fdc3d84132R2) to exclude them

Comment (Kehrlann) [reply metadata/org.testcontainers/testcontainers/1.19.8/reflect-config.json:1]: While I do not know the specifics of the type entries, when we compare what's currently in the repository for testcontainers (1.17.6), and diff with the new config, we get 8 new types and 1 type removed. Existing classes get many, many more conditions though, explaining the number of entries going from 358 to 452. So while the file is big, it's not adding many types to the resulting image. Details below.

---

Here are the 8 new types:

```
com.github.dockerjava.api.model.LoadResponseItem
com.github.dockerjava.api.model.RuntimeInfo
java.util.concurrent.ForkJoinTask
java.util.concurrent.atomic.AtomicMarkableReference
org.testcontainers.shaded.com.github.dockerjava.core.DockerContextMetaFile
org.testcontainers.shaded.com.github.dockerjava.core.DockerContextMetaFile$Endpoints
org.testcontainers.shaded.com.github.dockerjava.core.DockerContextMetaFile$Endpoints$Docker
sun.security.provider.SHA2$SHA256
```

And the 1 type removed, `[Lorg.testcontainers.shaded.com.google.common.collect.ImmutableMapEntry;` . 


We do have many more conditions per type, though. For example, `com.github.dockerjava.api.model.AuthConfig` went from 1.17.6 :

```json
{
  "condition": {
    "typeReachable": "org.testcontainers.DockerClientFactory"
  },
  "name": "com.github.dockerjava.api.model.AuthConfig",
  "allDeclaredFields": true,
  "allDeclaredConstructors": true
}
```

to 1.19.7:

```json
{
  "condition": {
    "typeReachable": "org.testcontainers.DockerClientFactory"
  },
  "name": "com.github.dockerjava.api.model.AuthConfig",
  "allDeclaredFields": true,
  "allDeclaredConstructors": true
}
{
  "name": "com.github.dockerjava.api.model.AuthConfig",
  "condition": {
    "typeReachable": "org.testcontainers.containers.GenericContainer"
  }
}
{
  "name": "com.github.dockerjava.api.model.AuthConfig",
  "condition": {
    "typeReachable": "org.testcontainers.containers.GenericContainer$$Lambda$478/0x000000e80123bd50"
  }
}
{
  "name": "com.github.dockerjava.api.model.AuthConfig",
  "queryAllDeclaredMethods": true,
  "condition": {
    "typeReachable": "org.testcontainers.dockerclient.DockerClientProviderStrategy"
  },
  "queryAllDeclaredConstructors": true
}
{
  "name": "com.github.dockerjava.api.model.AuthConfig",
  "condition": {
    "typeReachable": "org.testcontainers.dockerclient.EnvironmentAndSystemPropertyClientProviderStrategy"
  }
}
{
  "name": "com.github.dockerjava.api.model.AuthConfig",
  "condition": {
    "typeReachable": "org.testcontainers.dockerclient.TestcontainersHostPropertyClientProviderStrategy"
  }
}
{
  "name": "com.github.dockerjava.api.model.AuthConfig",
  "condition": {
    "typeReachable": "org.testcontainers.shaded.com.fasterxml.jackson.databind.ObjectMapper"
  },
  "allDeclaredFields": true,
  "methods": [
    {
      "name": "getAuth",
      "parameterTypes": []
    },
    {
      "name": "getEmail",
      "parameterTypes": []
    },
    {
      "name": "getIdentitytoken",
      "parameterTypes": []
    },
    {
      "name": "getPassword",
      "parameterTypes": []
    },
    {
      "name": "getRegistryAddress",
      "parameterTypes": []
    },
    {
      "name": "getRegistrytoken",
      "parameterTypes": []
    },
    {
      "name": "getStackOrchestrator",
      "parameterTypes": []
    },
    {
      "name": "getUsername",
      "parameterTypes": []
    }
  ]
}
{
  "name": "com.github.dockerjava.api.model.AuthConfig",
  "condition": {
    "typeReachable": "org.testcontainers.shaded.com.github.dockerjava.core.DefaultDockerClientConfig$Builder"
  }
}
{
  "name": "com.github.dockerjava.api.model.AuthConfig",
  "condition": {
    "typeReachable": "org.testcontainers.shaded.com.github.dockerjava.core.DockerConfigFile"
  },
  "allDeclaredFields": true,
  "methods": [
    {
      "name": "<init>",
      "parameterTypes": []
    }
  ]
}
{
  "name": "com.github.dockerjava.api.model.AuthConfig",
  "condition": {
    "typeReachable": "org.testcontainers.shaded.com.github.dockerjava.core.command.CreateContainerCmdImpl"
  }
}
{
  "name": "com.github.dockerjava.api.model.AuthConfig",
  "condition": {
    "typeReachable": "org.testcontainers.shaded.com.github.dockerjava.core.exec.AbstrDockerCmdExec"
  }
}
{
  "name": "com.github.dockerjava.api.model.AuthConfig",
  "condition": {
    "typeReachable": "org.testcontainers.shaded.com.github.dockerjava.core.exec.CreateContainerCmdExec"
  }
```

I don't see myself combing through all the new configs manually - and I'm not familiar with Testcontainers internals to begin with.

File: tests/src/org.testcontainers/testcontainers/1.19.8/build.gradle (hunk 2 above)
Comment (Kehrlann) [reply tests/src/org.testcontainers/testcontainers/1.19.8/build.gradle]: Copied the `build.gradle` as-is from from Testcontainers tests 1.17.6.

However it seems to works without JNA here so I removed the entry.

File: metadata/org.testcontainers/testcontainers/1.19.8/resource-config.json (hunk 1 above)
Comment (Kehrlann) [reply metadata/org.testcontainers/testcontainers/1.19.8/resource-config.json]: I just used the output of the agent as-is.

I tried without the new entries and it works just as well, thanks for the catch!

File: tests/src/org.testcontainers/testcontainers/1.19.8/src/main/java/org_testcontainers/testcontainers/FindJsonSerialization.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.reflect.Modifier;
+import java.nio.charset.StandardCharsets;
+import java.util.Set;
+
+import static org.reflections.scanners.Scanners.SubTypes;
+
+class FindJsonSerialization {
```
Comment (dnestoro) [inline tests/src/org.testcontainers/testcontainers/1.19.8/src/main/java/org_testcontainers/testcontainers/FindJsonSerialization.java:21]: Why did you need this way of collecting metadata? Couldn't the agent collect this metadata on a proper test?

Comment (Kehrlann) [reply tests/src/org.testcontainers/testcontainers/1.19.8/src/main/java/org_testcontainers/testcontainers/FindJsonSerialization.java:21]: I just re-applied what they did in the tests for the previous versions. 

My interpretation is the following: the agent does not catch any of those APIs, because the tests do not exercise these classes. The testcontainers API, and the underlying docker APIs, is huge. For example, the core testcontainers configuration interface, `Container`, has around 60 methods. It's most common implementation has 114 public methods. The methods, in turn, often use their own types (`Network`, `DockerClient`, ...). While there is redundancy in those methods, it's still a lot of surface to cover in tests. So, the initial implementation was done with a rather scattershot method that brings in the full - probably too much, but this being **test**containers, maybe the thinking is "test binaries are not as critical as production binaries" ?

To do this properly, we would have to spend quite a lot of time to nail down the correct tests, and maybe bring the Testcontainers team on board. I'm happy on refining the tests later on, for a later version, but I don't have the bandwidth right now.
